src/game.rs

```diff
@@ -21,90 +21,248 @@ pub enum ChessError {
+    }
+
+    fn piece_can_snipe(&self, at: &Position) -> bool {
+        self.get_piece(at).unwrap().can_snipe()
+    }
+
+    fn get_info_for_turn(&self, turn: usize) -> &(BoardState, Move) {
+        self.history.get_info_for_turn(turn)
+    }
+
+    fn has_castling_right(&self, right: CastlingRightsKind) -> bool {
+        self.get_current_state()
+            .castling_rights
+            .has_castling_right(right)
+    }
+
+    pub fn go_back_a_turn(&mut self) {
+        self.status.update(GameStatus::Replay);
+        self.history.previous_state()
+    }
+
+    pub fn go_forward_a_turn(&mut self) {
+        if self.status == GameStatus::Replay {
+            self.history.next_state();
+            if self.history.is_ongoing() {
+                self.status.update(GameStatus::Ongoing)
+            }
+        }
+    }
+
+    pub fn go_to_beginning(&mut self) {
+        self.status.update(GameStatus::Replay);
+        self.history.initial_state()
     }
 
-    pub fn get_board_state(&self) -> &BoardState {
-        &self.state
+    pub fn resume(&mut self) {
+        self.status.update(GameStatus::Ongoing);
+        self.history.resume()
+    }
+
+    fn add_info(&mut self, next_state: BoardState, mv: Move) {
+        self.history.add_info(next_state, mv);
     }
 
     pub fn move_piece(&mut self, from: Position, to: Position) -> ChessResult {
-        if let Some(piece) = self.state.get_piece(&from) {
-            let mv = Move::new(from, to);
+        if self.status != GameStatus::Replay {
+            if let Some(piece) = self.get_piece(&from) {
+                let mv = Move::new(from, to);
+                self.is_move_valid(&mv)?;
 
-            self.is_move_valid(&mv)?;
-            self.state.move_piece(&mv);
-            self.add_moves();
-            self.remove_self_checks();
-            self.update_status();
+                let mut next_state = self.clone_current_state();
+                next_state.move_piece(&mv);
+                self.history.add_info(next_state, mv);
+                self.update();
 
-            println!("{} : {}", piece, mv);
+                println!("{} : {}", piece, mv);
+            }
         }
         Ok(())
     }
 
+    fn update(&mut self) {
+        self.add_moves();
+        self.remove_self_checks();
+        self.update_status()
+    }
+
     fn update_status(&mut self) {
-        let mut next_turn = self.clone();
-        next_turn.state.player = !next_turn.state.player;
-        next_turn.add_moves();
-        if next_turn.has_check() {
-            if self.valid_moves.is_empty() {
-                self.status = GameStatus::Checkmate;
-            } else {
-                self.status = GameStatus::Check;
-            }
-        } else if self.valid_moves.is_empty() {
-            self.status = GameStatus::Stalemate;
-        } else {
-            self.status = GameStatus::Ongoing;
+        if self.status == GameStatus::Replay && self.history.is_ongoing() {
```

Comment:
We still need to check for the other statuses here. I think we can just say `if !self.history.is_ongoing()` .. set Replay and return.